pennylane/compiler/python_compiler/transforms/measurements_from_samples.py

```diff
@@ -0,0 +1,651 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains the implementation of the measurements_from_samples transform,
+written using xDSL."""
+
+from abc import abstractmethod
+from dataclasses import dataclass
+
+import jax
+import jax.numpy as jnp
+from xdsl import context, ir, passes, pattern_rewriter
+from xdsl.dialects import arith, builtin, func, tensor
+from xdsl.pattern_rewriter import PatternRewriter, RewritePattern
+from xdsl.rewriter import InsertPoint
+
+from pennylane.compiler.compiler import CompileError
+from pennylane.compiler.python_compiler import quantum_dialect as quantum
+from pennylane.compiler.python_compiler.jax_utils import xdsl_module
+from pennylane.compiler.python_compiler.transforms.api import xdsl_transform
+
+
+@dataclass(frozen=True)
+class MeasurementsFromSamplesPass(passes.ModulePass):
+    """Pass that replaces all terminal measurements in a program with a single
+    :func:`pennylane.sample` measurement, and adds postprocessing instructions to recover the
+    original measurement.
+    """
+
+    name = "measurements-from-samples"
+
+    # pylint: disable=arguments-renamed
+    def apply(self, _ctx: context.Context, module: builtin.ModuleOp) -> None:
+        """Apply the measurements-from-samples pass."""
+        shots = _get_static_shots_value_from_first_device_op(module)
+
+        # TODO: Do we want a walker or a greedy applier? TBD
+        # walker = pattern_rewriter.PatternRewriteWalker(MeasurementsFromSamplesPattern(shots))
+        # walker.rewrite_module(module)
+
+        greedy_applier = pattern_rewriter.GreedyRewritePatternApplier(
+            [
+                SamplePattern(shots),
+                ExpvalPattern(shots),
+                VarPattern(shots),
+                ProbsPattern(shots),
+                CountsPattern(shots),
+                StatePattern(shots),
+            ]
+        )
+        walker = pattern_rewriter.PatternRewriteWalker(greedy_applier)
```

Comment:
```suggestion
        walker = pattern_rewriter.PatternRewriteWalker(greedy_applier, apply_recursively=False)
```

With apply_recursively=False your patterns won't match against newly created operations. I believe your patterns don't generate ops that will need to be changed. So this may speed things up.

Regarding not using the GreedyRewritePatternApplier I think you need it as it essentially combines all patterns into one. I believe if you only had one pattern you may not need the GreedyRewritePatternApplier.

Optionally, one could write their own PatternApplier.

Reply:
c6ba5f644